freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UmsUsersStateProvider.java

```diff
@@ -217,11 +219,26 @@ private void addMemberToInternalTrackingGroup(UsersState.Builder usersStateBuild
         usersStateBuilder.addMemberToGroup(UserSyncConstants.CDP_USERSYNC_INTERNAL_GROUP, username);
     }
 
+    private Optional<String> getOptionalAzureObjectId(List<CloudIdentity> cloudIdentities) {
+        List<CloudIdentity> azureCloudIdentities = cloudIdentities.stream()
+                .filter(cloudIdentity -> cloudIdentity.getCloudIdentityDomain().hasAzureCloudIdentityDomain())
+                .collect(Collectors.toList());
+        if (azureCloudIdentities.isEmpty()) {
+            return Optional.empty();
+        }
+        if (azureCloudIdentities.size() > 1) {
+            throw new IllegalStateException(String.format("List contains multiple azure cloud identities = %s", cloudIdentities));
+        }
+        String azureObjectId = Iterables.getOnlyElement(azureCloudIdentities).getCloudIdentityName().getAzureCloudIdentityName().getObjectId();
+        return Optional.of(azureObjectId);
+    }
+
     private FmsUser umsUserToUser(User umsUser) {
         FmsUser fmsUser = new FmsUser();
         fmsUser.withName(umsUser.getWorkloadUsername());
         fmsUser.withFirstName(getOrDefault(umsUser.getFirstName(), "None"));
         fmsUser.withLastName(getOrDefault(umsUser.getLastName(), "None"));
+        fmsUser.withAzureObjectId(getOptionalAzureObjectId(umsUser.getCloudIdentitiesList()));
```

Comment:
The AzureObjectId belongs in the UmsUsersState, not the UsersState because it's information that is provided by the UMS with no matching information from the FreeIPA side. We use equality to compare the FmsUser and FmsGroup models generated by the UMS against those generated by FreeIPA. Adding in asymmetric information will break the equality and the state difference calculation.
see how we handle workload credentials instead. I think that you can pass the List<CloudIdentity> into the handleUser method and store it there. This will also let you only store OIDs for users that have access to the environment instead of all users.

Reply:
Thanks, I've moved cloud identify information into the UmsUserState.